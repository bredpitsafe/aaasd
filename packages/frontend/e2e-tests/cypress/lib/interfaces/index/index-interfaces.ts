export enum EPagesName {
    dashboard = 'Dashboard',
    tradingServersManager = 'Trading Servers Manager',
    herodotusTerminal = 'Herodotus Terminal',
    herodotusTrades = 'Herodotus Trades',
    herodotusDashboard = 'Herodotus Dashboard',
    tradingStats = 'Trading Stats',
    backtestingManager = 'Backtesting Manager',
    jsonCharts = 'JSON Charts',
    balanceMonitor = 'Balance Monitor',
    wsQueryTerminal = 'WS Query Terminal',
}
