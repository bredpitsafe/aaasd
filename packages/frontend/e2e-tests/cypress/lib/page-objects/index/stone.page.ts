import { BasePage } from '../../base.page';
import { Button } from '../../base/elements/button';
import { Text } from '../../base/elements/text';
enum EStonePageSelectors {
    Dashboard = '[href*=dashboard]',
    TradingServersManager = '[href*=trading-servers-manager]',
    HerodotusTerminal = '[href*=herodotus-terminal]',
    TradingStats = '[href*=trading-stats]',
    BacktestingManager = '[href*=backtesting]',
    BalanceMonitor = '[href*=balance-monitor]',
    WSQueryTerminal = '[href*=ws-query-terminal]',
    JsonCharts = '[href*=charter-json-viewer]',
    LoadingIcon = '[class*="anticon anticon-loading"]',
}

class StonePage extends BasePage {
    readonly dashboard = new Text(EStonePageSelectors.Dashboard, false);
    readonly tradingServersManager = new Text(EStonePageSelectors.TradingServersManager, false);
    readonly herodotusTerminal = new Text(EStonePageSelectors.HerodotusTerminal, false);
    readonly tradingStats = new Text(EStonePageSelectors.TradingStats, false);
    readonly backtestingManager = new Text(EStonePageSelectors.BacktestingManager, false);
    readonly balanceMonitor = new Text(EStonePageSelectors.BalanceMonitor, false);
    readonly wsQueryTerminal = new Text(EStonePageSelectors.WSQueryTerminal, false);
    readonly jsonCharts = new Text(EStonePageSelectors.JsonCharts, false);
    readonly loadingIcon = new Button(EStonePageSelectors.LoadingIcon, false);

    checkElementsExists(): void {
        this.dashboard.checkExists();
        this.tradingServersManager.checkExists();
        this.herodotusTerminal.checkExists();
        this.tradingStats.checkExists();
        this.backtestingManager.checkExists();
        this.backtestingManager.checkExists();
    }

    open(): StonePage {
        this.visit();
        return this;
    }
}

export const stonePage = new StonePage();
