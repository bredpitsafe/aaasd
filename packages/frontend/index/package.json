{
  "name": "@frontend/index",
  "version": "1.0.0",
  "description": "simple index page for choose application",
  "scripts": {
    "build": "NODE_ENV=production npx webpack",
    "lint": "eslint './src/**/*.{js,ts,tsx}' --quiet",
    "server": "npx webpack serve",
    "test_unit": "jest --passWithNoTests",
    "types": "tsc --noEmit"
  },
  "dependencies": {
    "@frontend/common": "^1.0.1",
    "@common/types": "^1.0.0"
  }
}
