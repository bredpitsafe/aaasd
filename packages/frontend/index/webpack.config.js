const {
    buildConfig,
    htmlConfig,
    defaultInOut,
    defaultDevServerConfig,
} = require('../../../webpack/configs');
const path = require('path');

const app = 'index';
const resolvePath = () => path.resolve(__dirname);

module.exports = buildConfig([
    defaultInOut([app], resolvePath),
    htmlConfig([app], resolvePath),
    defaultDevServerConfig(app),
]);
