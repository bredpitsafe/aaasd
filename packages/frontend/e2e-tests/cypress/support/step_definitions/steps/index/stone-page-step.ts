import { Given } from '@badeball/cypress-cucumber-preprocessor';

import { EPagesName } from '../../../../lib/interfaces/index/index-interfaces';
import { ERobots } from '../../../../lib/interfaces/trading-servers-manager/robots-interfaces';
import { backtestingPage } from '../../../../lib/page-objects/backtesting/backtesting.page';
import { balanceMonitorPage } from '../../../../lib/page-objects/balance-monitor/balance-monitor.page';
import { charterJsonViewerPage } from '../../../../lib/page-objects/charter-json-viewer/charter-json-viewer.page';
import { authenticationPage } from '../../../../lib/page-objects/common/authentication.page';
import { settingsModal } from '../../../../lib/page-objects/common/settings.modal';
import { tableRow } from '../../../../lib/page-objects/common/table/table.row';
import { dashboardPage } from '../../../../lib/page-objects/dashboard/dashboard.page';
import { herodotusTerminalPage } from '../../../../lib/page-objects/herodotus-terminal/herodotus-terminal.page';
import { herodotusTradesPage } from '../../../../lib/page-objects/herodotus-trades/herodotus-trades.page';
import { stonePage } from '../../../../lib/page-objects/index/stone.page';
import { tradingServersManagerPage } from '../../../../lib/page-objects/trading-servers-manager/trading-servers-manager.page';
import { tradingStatsPage } from '../../../../lib/page-objects/trading-stats/trading-stats.page';
import { wsQueryTerminalPage } from '../../../../lib/page-objects/ws-query-terminal/ws-query-terminal.page';
import { customWait } from '../../../../lib/web-socket/server';
import { checkUrlInclude } from '../../../asserts/comon-url-assert';

Given(`user goes to the "Stone" page`, () => {
    stonePage.visit();
    stonePage.checkElementsExists();
});

Given(`user click an the {string} link`, (pageName: EPagesName) => {
    switch (pageName) {
        case EPagesName.dashboard:
            stonePage.dashboard.click();
            break;
        case EPagesName.tradingServersManager:
            stonePage.tradingServersManager.click();
            break;
        case EPagesName.herodotusTerminal:
            stonePage.herodotusTerminal.click();
            break;
        case EPagesName.tradingStats:
            stonePage.tradingStats.click();
            break;
        case EPagesName.backtestingManager:
            stonePage.backtestingManager.click();
            break;
        case EPagesName.jsonCharts:
            stonePage.jsonCharts.click();
            break;
        case EPagesName.balanceMonitor:
            stonePage.balanceMonitor.click();
            break;
        case EPagesName.wsQueryTerminal:
            stonePage.wsQueryTerminal.click();
            break;
    }
});

Given(
    `user on the {string} page with the "default" backend server parameter`,
    (pageName: EPagesName) => {
        switch (pageName) {
            case EPagesName.tradingServersManager:
                tradingServersManagerPage.checkElementsExists();
                break;
            case EPagesName.herodotusTerminal:
                herodotusTerminalPage.checkElementsExists();
                break;
            case EPagesName.tradingStats:
                tradingStatsPage.checkElementsExists();
                break;
            case EPagesName.backtestingManager:
                backtestingPage.checkElementsExists();
                break;
            case EPagesName.balanceMonitor:
                balanceMonitorPage.checkElementsExists();
                break;
        }
    },
);

Given(`user not sees data on the {string} page`, (pageName: EPagesName) => {
    switch (pageName) {
        case EPagesName.tradingServersManager:
            tradingServersManagerPage.checkNotVisibleTabPanel();
            tradingServersManagerPage.checkVisibleLoading();
            break;
        case EPagesName.herodotusTerminal:
            herodotusTerminalPage.checkVisibleTabPanel();
            herodotusTerminalPage.checkTaskTableVisibility(false);
            stonePage.loadingIcon.checkVisible();
            break;
        case EPagesName.backtestingManager:
            backtestingPage.checkNotExistsTask();
            break;
    }
});

Given(`user sees the {string} page`, (pageName: EPagesName) => {
    switch (pageName) {
        case EPagesName.tradingServersManager:
            tradingServersManagerPage.checkVisibleTabPanel();
            checkUrlInclude(tradingServersManagerPage.pageUrl);
            break;
        case EPagesName.herodotusTerminal:
            herodotusTerminalPage.checkVisibleTabPanel();
            herodotusTerminalPage.checkTaskTableVisibility(true);
            checkUrlInclude(herodotusTerminalPage.pageUrl);
            break;
        case EPagesName.tradingStats:
            tradingStatsPage.checkDailyVisibleTable();
            checkUrlInclude(tradingStatsPage.pageUrl);
            break;
        case EPagesName.backtestingManager:
            backtestingPage.checkVisiblePanel();
            checkUrlInclude(backtestingPage.pageUrl);
            break;
        case EPagesName.balanceMonitor:
            balanceMonitorPage.checkVisiblePanel();
            checkUrlInclude(balanceMonitorPage.pageUrl);
            break;
        case EPagesName.dashboard:
            dashboardPage.checkElementsExists();
            checkUrlInclude(dashboardPage.pageUrl);
            break;
        case EPagesName.jsonCharts:
            charterJsonViewerPage.checkElementsExists();
            checkUrlInclude(charterJsonViewerPage.pageUrl);
            break;
        case EPagesName.herodotusTrades:
            herodotusTradesPage.checkElementsExists();
            herodotusTradesPage.checkVisibleTable();
            checkUrlInclude(herodotusTradesPage.pageUrl);
            break;
        case EPagesName.wsQueryTerminal:
            wsQueryTerminalPage.checkElementsExists();
            wsQueryTerminalPage.checkVisibleTab();
            checkUrlInclude(wsQueryTerminalPage.pageUrl);
            break;
    }
});

Given(
    `user goes to the {string} page with the backend server parameter`,
    (pageName: EPagesName) => {
        switch (pageName) {
            case EPagesName.tradingServersManager:
                tradingServersManagerPage.openPageWithBackendServer();
                break;
            case EPagesName.herodotusTerminal:
                herodotusTerminalPage.openPageWithBackendServer();
                break;
            case EPagesName.tradingStats:
                tradingStatsPage.openPageWithBackendServer();
                break;
            case EPagesName.backtestingManager:
                backtestingPage.openPageWithBackendServer();
                break;
            case EPagesName.balanceMonitor:
                balanceMonitorPage.openPageWithBackendServer();
                break;
            case EPagesName.wsQueryTerminal:
                wsQueryTerminalPage.openPageWithBackendServer();
                break;
        }
    },
);

Given(
    `user goes to the {string} page by {string} server params`,
    (pageName: EPagesName, nameServer: string) => {
        switch (pageName) {
            case EPagesName.tradingServersManager:
                tradingServersManagerPage.openPageByBackendServer(nameServer);
                break;
            case EPagesName.herodotusTerminal:
                herodotusTerminalPage.openPageByBackendServer(nameServer);
                break;
            case EPagesName.tradingStats:
                tradingStatsPage.openPageByBackendServer(nameServer);
                break;
            case EPagesName.backtestingManager:
                backtestingPage.openPageByBackendServer(nameServer);
                break;
            case EPagesName.balanceMonitor:
                balanceMonitorPage.openPageByBackendServer(nameServer);
                break;
        }
    },
);

Given(
    `user goes to the "Backtesting Manager" page with the backend server {string}`,
    (nameServer: string) => {
        backtestingPage.openPageWithBackendServerByParams(nameServer);
    },
);

Given(`user goes to the "Herodotus Terminal" page by index {string}`, (index: string) => {
    herodotusTerminalPage.openPageWithBackendServerByIndex(index);
});

Given(`user goes to the "Herodotus Terminal" page by name {string}`, (robotName: string) => {
    const index = `${ERobots[robotName]}`;
    herodotusTerminalPage.openPageWithBackendServerByIndex(index);
});

Given(`user click "Log Out" button on the {string} page`, () => {
    authenticationPage.clickLogOutButton();
});

Given(`user on the {string} page with the "ModalSettings" modal`, (pageName: EPagesName) => {
    settingsModal.checkElementsExists();
    customWait(1);
    switch (pageName) {
        case EPagesName.tradingServersManager:
            checkUrlInclude(tradingServersManagerPage.pageUrl);
            break;
        case EPagesName.herodotusTerminal:
            checkUrlInclude(herodotusTerminalPage.pageUrl);
            break;
        case EPagesName.tradingStats:
            checkUrlInclude(tradingStatsPage.pageUrl);
            break;
        case EPagesName.backtestingManager:
            checkUrlInclude(backtestingPage.pageUrl);
            break;
    }
});

Given(`user opens the {string} page`, (pageName: EPagesName) => {
    switch (pageName) {
        case EPagesName.dashboard:
            dashboardPage.visit();
            dashboardPage.checkElementsExists();
            break;
        case EPagesName.tradingServersManager:
            tradingServersManagerPage.visit();
            settingsModal.checkElementsExists();
            break;
        case EPagesName.herodotusTerminal:
            herodotusTerminalPage.visit();
            settingsModal.checkElementsExists();
            break;
        case EPagesName.tradingStats:
            tradingStatsPage.visit();
            settingsModal.checkElementsExists();
            break;
        case EPagesName.backtestingManager:
            backtestingPage.visit();
            settingsModal.checkElementsExists();
            break;
        case EPagesName.balanceMonitor:
            balanceMonitorPage.visit();
            settingsModal.checkElementsExists();
            break;
        case EPagesName.wsQueryTerminal:
            wsQueryTerminalPage.visit();
            settingsModal.checkElementsExists();
            break;
    }
});

Given(`user on the "Herodotus Terminal" page of the {string} robot`, (robotName: ERobots) => {
    checkUrlInclude(herodotusTerminalPage.pageUrl);
    herodotusTerminalPage.checkElementsExists();
    checkUrlInclude(`${ERobots[robotName]}`);
});

Given(
    `user types the modal settings with the {string} backend server parameter`,
    (backendServerName: string) => {
        if (backendServerName === 'default') {
            backendServerName = Cypress.env('backendServerName');
        }
        settingsModal.serverSelect.checkVisible();
        customWait(0.5);
        settingsModal.serverSelect.typeAndClickByText(backendServerName);
    },
);

Given(`user gets URL the "Charts" page and opens the page`, () => {
    tableRow.visitPageByLinkInRow();
});
