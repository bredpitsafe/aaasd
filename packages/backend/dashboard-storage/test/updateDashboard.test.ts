import type { TDashboardItem } from '@grpc-schemas/dashboard_storage-api-sdk/services/dashboard_storage/v1/common';
import { pick } from 'lodash-es';
import { afterAll, beforeEach, describe, expect, it } from 'vitest';

import { Client, users } from './client/index.ts';
import initialDashboard from './fixtures/dashboard.json';
import { getDashboard } from './utils/index.js';

describe('UpdateDashboard', () => {
    const client = new Client({ authenticate: true });
    const otherClient = new Client({ authenticate: true, as: users.user2.username });

    let dashboard: TDashboardItem | undefined = undefined;
    let id = '';

    beforeEach(async () => {
        const createDashboardResponse = await client.request(initialDashboard);
        expect(createDashboardResponse.state).toBe('Done');
        expect(createDashboardResponse.payload.id).toMatch(/[a-f0-9-]+/);
        id = createDashboardResponse.payload.id;

        dashboard = await getDashboard(client, id);
    });

    afterAll(() => {
        client.close();
        otherClient.close();
    });

    it('update', async () => {
        const updatedDashboard = {
            type: 'UpdateDashboard',
            ...pick(dashboard, 'id', 'digest'),
            name: 'modified name',
            config: 'modified config',
            status: 'Suspended',
        };

        const response = await client.request(updatedDashboard);
        expect(response.state).toBe('Done');
        expect(response.payload.type).toBe('DashboardUpdated');

        // Check that dashboard name and status has been updated
        const newDashboard = await getDashboard(client, id);
        expect(newDashboard?.name).toBe(updatedDashboard.name);
        expect(newDashboard?.status).toBe(updatedDashboard.status);

        // Check that dashboard config has been updated
        const fetchedDashboard = await client.request({ type: 'FetchDashboardConfig', id });
        expect(fetchedDashboard.payload.config).toBe(updatedDashboard.config);
    });

    it('same digest update does succeeds and drops draft', async () => {
        // Step 1. Fetch config for dashboard.
        const {
            payload: { config },
        } = await client.request({ type: 'FetchDashboardConfig', id });

        // Step 2. Update draft with a same config value.
        const draftUpdateResponse = await client.request({
            type: 'UpdateDashboardDraft',
            id,
            config,
        });
        expect(draftUpdateResponse.payload.type).toBe('DashboardDraftUpdated');

        // Step 3. Update dashboard with a same config.
        const request = {
            type: 'UpdateDashboard',
            ...pick(dashboard, 'id', 'name', 'status', 'digest'),
            config,
        };
        const response = await client.request(request);
        expect(response.payload.type).toBe('DashboardUpdated');

        // Step 4. Check that draft doesn't exist anymore.
        const dashboardWithoutDraft = await getDashboard(client, id);
        expect(dashboardWithoutDraft?.hasDraft).toBe(false);
    });

    it('digest mismatch', async () => {
        const updatedDashboard = {
            type: 'UpdateDashboard',
            ...pick(dashboard, 'id'),
            name: 'modified name',
            config: 'modified config',
            status: 'Suspended',
            digest: 'definitely_incorrect_digest',
        };

        const response = await client.request(updatedDashboard);

        expect(response.error).toMatchSnapshot();
    });

    it('resets draft after update', async () => {
        // Step 1. Create draft.
        const draft = {
            ...pick(dashboard, 'id'),
            type: 'UpdateDashboardDraft',
            config: 'modified config draft',
        };
        const draftResponse = await client.request(draft);
        expect(draftResponse.state).toBe('Done');
        expect(draftResponse.payload.type).toBe('DashboardDraftUpdated');

        // Step 2. Update dashboard.
        const updatedDashboard = {
            type: 'UpdateDashboard',
            ...pick(dashboard, 'id', 'status', 'digest'),
            name: 'modified name',
            config: 'modified config',
        };
        const updateResponse = await client.request(updatedDashboard);
        expect(updateResponse.state).toBe('Done');
        expect(updateResponse.payload.type).toBe('DashboardUpdated');

        // Step 3. Check draft, it should be reset.
        const fetchResponse = await client.request({
            type: 'SubscribeToDashboard',
            id,
        });
        expect(fetchResponse.payload.dashboard.name).toBe(updatedDashboard.name);
        expect(fetchResponse.payload.dashboard.hasDraft).toBe(false);
    });

    it('other users`s dashboard', async () => {
        const updatedDashboard = {
            type: 'UpdateDashboard',
            ...pick(dashboard, 'id', 'digest'),
            name: 'modified name',
            config: 'modified config',
            status: 'Suspended',
        };
        const response = await otherClient.request(updatedDashboard);

        expect(response.error).toMatchSnapshot();
    });
});
