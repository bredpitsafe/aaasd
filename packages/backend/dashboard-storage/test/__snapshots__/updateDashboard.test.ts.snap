// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UpdateDashboard > digest mismatch 1`] = `
{
  "args": {
    "currentDigest": "3388ab53e11ad959493f443fd2da8856",
    "receivedDigest": "definitely_incorrect_digest",
  },
  "description": "Dashboard update failed: Digest mismatch. You are trying to update an outdated dashboard.",
  "kind": "Validation",
}
`;

exports[`UpdateDashboard > other users\`s dashboard 1`] = `
{
  "description": "Dashboard update failed: You do not have permission to edit this dashboard",
  "kind": "Authorization",
}
`;
